File: python/ch1_arrays_and_strings/1_1.py
"""
Cracking the Coding Interview Question 1.1 Is Unique

Implement an algorithm to determine if a string has all unique characters. What
if you cannot use additional data structures?
"""

import unittest


def is_unique(string):
    return len(string) == len(set(string))  # set doesn't allow duplicate chars

def is_unique_no_data_structure(string):  # trades speed for space
    length = len(string)
    for i in range(length - 1):
        for j in range(i + 1, length):
            if string[i] == string[j]:  # cannot be unique if two chars are equal
                return False

    return True

class Test(unittest.TestCase):
    data_expected_true = ('abcd', 'jose', 'rojas', 'world')
    data_expected_false = ('hello', 'good', 'buddy')

    def test_is_unique(self):
        for test_string in self.data_expected_true:
            actual = is_unique(test_string)
            self.assertTrue(actual)

        for test_string in self.data_expected_false:
            actual = is_unique(test_string)
            self.assertFalse(actual)

if __name__ == '__main__':
    unittest.main()

File: javascript/ch1_arrays_and_strings/1_1.js
/*
Cracking the Coding Interview Question 1.1 Is Unique

Implement an algorithm to determine if a string has all unique characters. What
if you cannot use additional data structures?
*/

var assert = require('chai').assert;

function isUnique(str) {
  let charSet = new Set();

  for (let char of str) {
    if (charSet.has(char)) {
      return false;
    }
    charSet.add(char);
  }
  return true;
}

function isUniqueNoDataStructues(str) { // trades speed for space
  let length = str.length;

  for (let i = 0; i < length - 1; i++) {
    for (let j = i + 1; j < length; j++) {
      if (str[i] == str[j]) {
        return false;
      }
    }
  }
  return true;
}

describe('isUnique', function() {
  let dataExpectedTrue = ['abcd', 'jose', 'rojas', ' '];
  let dataExpectedFalse = ['hello', 'good', 'buddy', '  '];

  for (let str of dataExpectedTrue) {
    it(`${str} is unique`, function() {
      assert.isTrue(isUniqueNoDataStructues(str));
    });
  }

  for (let str of dataExpectedFalse) {
    it(`${str} is not unique`, function() {
      assert.isFalse(isUniqueNoDataStructues(str));
    });
  }
});
